rhp/v4/rpc.go

```diff
@@ -0,0 +1,529 @@
+type ContractRevision struct {
+	ID       types.FileContractID `json:"id"`
+	Revision types.V2FileContract `json:"revision"`
+}
+
+// RPCSettings returns the current settings of the host
+func RPCSettings(t TransportClient) (rhp4.HostSettings, error) {
+	s := t.DialStream()
+	defer s.Close()
+
+	if err := rhp4.WriteRequest(s, rhp4.RPCSettingsID, nil); err != nil {
+		return rhp4.HostSettings{}, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	var resp rhp4.RPCSettingsResponse
+	if err := rhp4.ReadResponse(s, &resp); err != nil {
+		return rhp4.HostSettings{}, fmt.Errorf("failed to read response: %w", err)
+	}
+	return resp.Settings, nil
+}
+
+// RPCReadSector reads a sector from the host
+func RPCReadSector(t TransportClient, prices rhp4.HostPrices, token rhp4.AccountToken, root types.Hash256, offset, length uint64) ([]byte, error) {
+	if offset+length > rhp4.SectorSize {
+		return nil, fmt.Errorf("read exceeds sector bounds")
+	}
+
+	s := t.DialStream()
+	defer s.Close()
+
+	if err := rhp4.WriteRequest(s, rhp4.RPCReadSectorID, &rhp4.RPCReadSectorRequest{
+		Prices: prices,
+		Token:  token,
+		Root:   root,
+		Offset: offset,
+		Length: length,
+	}); err != nil {
+		return nil, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	var resp rhp4.RPCReadSectorResponse
+	if err := rhp4.ReadResponse(s, &resp); err != nil {
+		return nil, fmt.Errorf("failed to read response: %w", err)
+	}
+	// TODO: verify proof + stream to writer
+	return resp.Sector, nil
+}
+
+// RPCWriteSector writes a sector to the host
+func RPCWriteSector(t TransportClient, prices rhp4.HostPrices, token rhp4.AccountToken, data []byte, duration uint64) (types.Hash256, error) {
+	if len(data) > rhp4.SectorSize {
+		return types.Hash256{}, fmt.Errorf("sector must be less than %d bytes", rhp4.SectorSize)
+	} else if len(data)%rhp4.LeafSize != 0 {
+		return types.Hash256{}, fmt.Errorf("sector must be a multiple of %d bytes", rhp4.LeafSize)
+	}
+
+	s := t.DialStream()
+	defer s.Close()
+
+	if err := rhp4.WriteRequest(s, rhp4.RPCWriteSectorID, &rhp4.RPCWriteSectorRequest{
+		Prices:   prices,
+		Token:    token,
+		Sector:   data,
+		Duration: duration,
+	}); err != nil {
+		return types.Hash256{}, fmt.Errorf("failed to write request: %w", err)
+	}
+
+	var resp rhp4.RPCWriteSectorResponse
+	if err := rhp4.ReadResponse(s, &resp); err != nil {
+		return types.Hash256{}, fmt.Errorf("failed to read response: %w", err)
+	}
+
+	// TODO: stream?
+	var sector [rhp4.SectorSize]byte
+	copy(sector[:], data)
+	root := rhp4.SectorRoot(&sector)
+	if root != resp.Root {
+		return types.Hash256{}, fmt.Errorf("invalid root returned: expected %v, got %v", root, resp.Root)
+	}
+	return root, nil
+}
+
+// RPCModifySectors modifies sectors on the host
+func RPCModifySectors(t TransportClient, cs consensus.State, prices rhp4.HostPrices, sk types.PrivateKey, contract ContractRevision, actions []rhp4.WriteAction) (types.V2FileContract, error) {
```

Comment:
Should we consider adjusting the RPC on the host side to return the amount spent? The RPC are going to be calling `prices.RPC*Price`, which seems like it could be done outside of the RPC function easily enough and we can ding the host if it doesn't match up.